File: 12. Error Handling and Building Fault Tolerant Systems/code/typescript/go_config_validation_at_boot_1.ts
// TypeScript, Config validation at boot
export interface Config {
    DatabaseURL: string;
    OpenAIKey: string;
    JWTSecret: string;
    ResendAPIKey: string;
}

// mustLoad throws if any required variable is missing.
// Call this once in main() before starting the server.
export function mustLoad(): Config {
    const required = [
        "DATABASE_URL",
        "OPENAI_API_KEY",
        "JWT_SECRET",
        "RESEND_API_KEY",
    ];

    const missing = required.filter(key => !process.env[key]);
    if (missing.length > 0) {
        // Crash immediately, loud and clear
        throw new Error(`[FATAL] missing required env vars: ${missing.join(", ")}`);
    }

    return {
        DatabaseURL: process.env.DATABASE_URL!,
        OpenAIKey: process.env.OPENAI_API_KEY!,
        JWTSecret: process.env.JWT_SECRET!,
        ResendAPIKey: process.env.RESEND_API_KEY!,
    };
}

// main.ts
if (require.main === module) {
    const cfg = mustLoad(); // throws here if config invalid
    // const server = newServer(cfg);
    // server.listen();
}

File: 12. Error Handling and Building Fault Tolerant Systems/code/rust/go_config_validation_at_boot_1.rs
// Rust, Config validation at boot
use std::env;

pub struct Config {
    pub database_url: String,
    pub openai_key: String,
    pub jwt_secret: String,
    pub resend_api_key: String,
}

// must_load panics if any required variable is missing.
// Call this once in main() before starting the server.
pub fn must_load() -> Config {
    let required = [
        "DATABASE_URL",
        "OPENAI_API_KEY",
        "JWT_SECRET",
        "RESEND_API_KEY",
    ];

    let missing: Vec<&str> = required.into_iter()
        .filter(|&key| env::var(key).is_err())
        .collect();

    if !missing.is_empty() {
        // Crash immediately, loud and clear
        panic!("[FATAL] missing required env vars: {}", missing.join(", "));
    }

    Config {
        database_url: env::var("DATABASE_URL").unwrap(),
        openai_key: env::var("OPENAI_API_KEY").unwrap(),
        jwt_secret: env::var("JWT_SECRET").unwrap(),
        resend_api_key: env::var("RESEND_API_KEY").unwrap(),
    }
}

// main.rs
fn main() {
    let _cfg = must_load();  // panics here if config invalid
    // let server = new_server(cfg);
    // server.listen();
}
